plugins/inputs/t128_tank/reader.go

```diff
@@ -0,0 +1,354 @@
+	tankAddress string
+	// the target port for the TANK instance
+	tankPort int
+	// time to wait before restarting the tank read command
+	restartDelay time.Duration
+	// used to notify the main routine that the read routine finished with an error
+	readDone chan error
+	// used to mock tank read subprocess in testing
+	tankReadCmdCtx CommandContext
+	// path to the index file
+	indexPath string
+	// default index in case reading to/from the index file fails
+	defaultIndex index
+}
+
+func NewReader(tankAddress string, tankPort int, topic string, indexPath string, defaultIndex index) *Reader {
+	return &Reader{
+		topic:          topic,
+		tankReadCmdCtx: exec.CommandContext,
+		sendChan:       make(chan IndexedMessage),
+		readDone:       make(chan error),
+		restartDelay:   5 * time.Second,
+		tankAddress:    tankAddress,
+		tankPort:       tankPort,
+		indexPath:      indexPath,
+		defaultIndex:   defaultIndex,
+	}
+}
+
+func (r *Reader) withTankReadCommandContext(tankReadCmdCtx CommandContext) *Reader {
+	r.tankReadCmdCtx = tankReadCmdCtx
+	return r
+}
+
+func (r *Reader) Run(mainCtx context.Context, stopchan chan struct{}) {
+	readCtx, readCtxCancel := context.WithCancel(mainCtx)
+	paused := false
+	lastIndex, err := getIndex(r.indexPath, r.defaultIndex)
+	r.lastSavedIndex = lastIndex.value
+	if err != nil {
+		log.Printf("Error in get index")
+		return
+	}
+
+	nextSaveCheck := time.After(2 * time.Second)
+	defer func() {
+		readCtxCancel()
+		setIndex(r.indexPath, index{value: r.lastSavedIndex})
+	}()
+
+	go r.read(mainCtx, readCtx, lastIndex.next())
+
+	for {
+		select {
+		case <-mainCtx.Done():
+			fmt.Printf("%s reader done", r.topic)
+			readCtxCancel()
+			return
+		case <-nextSaveCheck:
+			if lastIndex.value > r.lastSavedIndex {
+				setIndex(r.indexPath, lastIndex)
+				r.lastSavedIndex = lastIndex.value
+			}
+			nextSaveCheck = time.After(2 * time.Second)
+		case err := <-r.readDone:
+			if !paused {
+				readCtx, readCtxCancel = context.WithCancel(mainCtx)
+				var boundaryFault *boundaryFault
+				if err != nil && errors.As(err, &boundaryFault) {
+					fmt.Printf("detected boundary fault, restarting %s tank read from index %d", r.topic, boundaryFault.nextAvailableIndex.value)
+					go r.read(mainCtx, readCtx, boundaryFault.nextAvailableIndex)
+				} else {
+					go r.read(mainCtx, readCtx, lastIndex.next())
+				}
+			} else {
+				fmt.Printf("reader is paused, skipping %s tank read restart", r.topic)
+			}
+		}
+	}
+}
+
+func (r *Reader) read(mainCtx context.Context, readCtx context.Context, startingIndex index) {
+	fmt.Printf("starting %s tank read from index %d", r.topic, startingIndex.value)
+	err := r.readFromTank(readCtx, startingIndex)
+	if err != nil {
+		fmt.Printf("%s read routine exited with error: %v", r.topic, err)
+	}
+
+	if mainCtx.Err() == nil {
+		time.Sleep(r.restartDelay)
+	}
+
+	r.readDone <- err
+}
+
+func (i *IndexedMessage) WithTimestamp(timestamp int) []byte {
```

Comment:
Is this needed?

Reply:
If we want to capture the timestamp. If not I can remove it from the struct also.

Reply:
We should not need to ensure a timestamp. Telegraf should use the current timestamp is one is not provided.